File: CursoIWTraining/Advanced/PersistenciaArquivo/app/src/main/java/com/dvidal/persistenciaarquivo/MainActivity.java
package com.dvidal.persistenciaarquivo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private EditText txtMessage;

    private Button btnRecord;
    private Button btnClear;
    private Button btnLoad;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Instâncias
        txtMessage = (EditText) findViewById(R.id.txtMessage);

        btnRecord = (Button) findViewById(R.id.btnRecord);
        btnClear = (Button) findViewById(R.id.btnClear);
        btnLoad = (Button) findViewById(R.id.btnLoad);

        // Listeners
        btnRecord.setOnClickListener(this);
        btnClear.setOnClickListener(this);
        btnLoad.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {

        switch (v.getId()){

            case R.id.btnRecord:
                record();
                break;

            case R.id.btnClear:
                clear();
                break;

            case R.id.btnLoad:
                load();
                break;
        }
    }

    private void record(){

        try {
            FileOutputStream output = openFileOutput("my_file.txt", MODE_PRIVATE);
            whiteFile(output);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private void clear(){

        txtMessage.setText("");
    }

    private void load(){


        try {
            FileInputStream input = openFileInput("my_file.txt");
            readFile(input);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void whiteFile(FileOutputStream fos) throws IOException{

        String[] lines = TextUtils.split(txtMessage.getText().toString(), "\n");
        PrintWriter writter = new PrintWriter(fos);
        for (String l: lines) {
            writter.println(l);
        }

        writter.flush(); // fecha processo
        writter.close(); // fecha o aquivo
        fos.close();     // fecha o output stream
    }

    private void readFile(FileInputStream fis) throws IOException{

        BufferedReader reader = new BufferedReader(new InputStreamReader(fis));
        StringBuilder sb = new StringBuilder();
        String line;

        while ((line = reader.readLine()) != null ){
            if (sb.length() != 0) sb.append("\n");
            sb.append(line);
        }

        reader.close();
        fis.close();

        txtMessage.setText(sb.toString());
    }
}

File: BroadcastReceiverTest/app/src/main/java/br/com/vp/broadcastreceivertest/MainActivity.kt
package br.com.vp.broadcastreceivertest

import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import br.com.vp.broadcastreceivertest.broadcasts.SecondBroadcastReceiver

class MainActivity : AppCompatActivity() {

    private val broadcastReceiver = SecondBroadcastReceiver()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        createIntentFilter()
    }

    fun updateScreen() {

        findViewById<View>(android.R.id.content).setBackgroundColor(Color.CYAN)
    }

    private fun createIntentFilter() {

        IntentFilter().also {
            with(it){
                addAction("BROADCAST_API")
                addCategory("BROADCAST_API_CATEGORY")
                registerReceiver(broadcastReceiver, it)
            }
        }
    }

    fun callBroadcastReceiver(v: View) {

        val label = (v as? Button)?.text.toString().toUpperCase()
        Intent(label).also {
            sendBroadcast(it)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(broadcastReceiver)
    }
}
